framework/src/meshgenerators/Poly2TriMeshGenerator.C

```diff
@@ -0,0 +1,245 @@
+  params.addParam<std::vector<MeshGeneratorName>>(
+      "holes", std::vector<MeshGeneratorName>(), "The MeshGenerators that define mesh holes.");
+  params.addParam<std::vector<bool>>(
+      "stitch_holes", std::vector<bool>(), "Whether to stitch to the mesh defining each hole.");
+  params.addParam<std::vector<bool>>("refine_holes",
+                                     std::vector<bool>(),
+                                     "Whether to allow automatically refining each hole boundary.");
+  params.addParam<Real>(
+      "desired_area", std::numeric_limits<Real>::max(), "Desired (maximum) triangle area.");
+  params.addParam<std::string>("desired_area_func",
+                               std::string(),
+                               "Function specifying desired triangle area as a function of x,y.");
+  params.addParam<MooseEnum>(
+      "algorithm",
+      algorithm,
+      "Control the use of binary search for the nodes of the stitched surfaces.");
+  params.addParam<bool>(
+      "verbose_stitching", false, "Whether mesh stitching should have verbose output.");
+
+  params.addClassDescription("Triangulates meshes within boundaries defined by input meshes.");
+
+  return params;
+}
+
+Poly2TriMeshGenerator::Poly2TriMeshGenerator(const InputParameters & parameters)
+  : MeshGenerator(parameters),
+    _bdy_ptr(getMesh("boundary")),
+    _add_nodes_per_boundary_segment(getParam<unsigned int>("add_nodes_per_boundary_segment")),
+    _refine_bdy(getParam<bool>("refine_boundary")),
+    _smooth_tri(getParam<bool>("smooth_triangulation")),
+    _hole_ptrs(getMeshes("holes")),
+    _stitch_holes(getParam<std::vector<bool>>("stitch_holes")),
+    _refine_holes(getParam<std::vector<bool>>("refine_holes")),
+    _desired_area(getParam<Real>("desired_area")),
+    _desired_area_func(getParam<std::string>("desired_area_func")),
+    _algorithm(parameters.get<MooseEnum>("algorithm")),
+    _verbose_stitching(parameters.get<bool>("verbose_stitching"))
+{
+  if (!_stitch_holes.empty() && _stitch_holes.size() != _hole_ptrs.size())
+    paramError("stitch_holes", "Need one stitch_holes entry per hole, if specified.");
+
+  for (auto hole_i : index_range(_stitch_holes))
+    if (hole_i < _refine_holes.size() && _refine_holes[hole_i])
+      paramError("refine_holes", "Disable auto refine of any hole boundary to be stitched.");
+}
+
+std::unique_ptr<MeshBase>
+Poly2TriMeshGenerator::generate()
+{
+  // We put the boundary mesh in a local pointer
+  std::unique_ptr<UnstructuredMesh> mesh = dynamic_pointer_cast<UnstructuredMesh>(_bdy_ptr);
+
+  Poly2TriTriangulator poly2tri(*mesh);
+  poly2tri.triangulation_type() = TriangulatorInterface::PSLG;
+
+  poly2tri.set_interpolate_boundary_points(_add_nodes_per_boundary_segment);
+  poly2tri.set_refine_boundary_allowed(_refine_bdy);
+
+  poly2tri.desired_area() = _desired_area;
+  poly2tri.minimum_angle() = 0; // Not yet supported
```

Comment:
any strong reason we dont support it?

Reply:
Because it'll take a long time to support.  It's a feature we can simply turn on with the Triangle interface, but Poly2Tri doesn't support refinement *at all* and I had to write refinement code from scratch.  Just handling all the corner cases from area-based refinement took quite a while.